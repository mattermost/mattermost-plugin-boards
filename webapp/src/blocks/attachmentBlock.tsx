// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Block, createBlock} from './block'

type AttachmentBlockFields = {
    fileId: string
}

type AttachmentBlock = Block & {
    type: 'attachment'
    fields: AttachmentBlockFields
    isUploading: boolean
    uploadingPercent: number
}

function createAttachmentBlock(block?: Block): AttachmentBlock {
    return {
        ...createBlock(block),
        type: 'attachment',
        fields: {
            fileId: block?.fields.attachmentId || block?.fields.fileId || '',
        },
        isUploading: false,
        uploadingPercent: 0,
    }
}

export {AttachmentBlock, createAttachmentBlock}
