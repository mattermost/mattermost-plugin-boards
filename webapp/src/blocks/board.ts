// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.


import difference from 'lodash/difference'

import {Utils, IDType} from '../utils'

import {Block, BlockPatch, createPatchesFromBlocks} from './block'
import {Card} from './card'

const BoardTypeOpen = 'O'
const BoardTypePrivate = 'P'
const boardTypes = [BoardTypeOpen, BoardTypePrivate]
type BoardTypes = typeof boardTypes[number]

enum MemberRole {
    Viewer = 'viewer',
    Commenter = 'commenter',
    Editor = 'editor',
    Admin = 'admin',
    None = '',
}

type Board = {
    id: string
    teamId: string
    channelId?: string
    createdBy: string
    modifiedBy: string
    type: BoardTypes
    minimumRole: MemberRole

    title: string
    description: string
    icon?: string
    showDescription: boolean
    isTemplate: boolean
    templateVersion: number
    properties: Record<string, string | string[]>
    cardProperties: IPropertyTemplate[]

    createAt: number
    updateAt: number
    deleteAt: number
}

type BoardPatch = {
    type?: BoardTypes
    minimumRole?: MemberRole
    title?: string
    description?: string
    icon?: string
    showDescription?: boolean
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    updatedProperties?: Record<string, any>
    deletedProperties?: string[]
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    updatedCardProperties?: IPropertyTemplate[]
    deletedCardProperties?: string[]
}

type BoardMember = {
    boardId: string
    userId: string
    roles?: string
    minimumRole: MemberRole
    schemeAdmin: boolean
    schemeEditor: boolean
    schemeCommenter: boolean
    schemeViewer: boolean
    synthetic: boolean
}

type BoardsAndBlocks = {
    boards: Board[]
    blocks: Block[]
}

type BoardsAndBlocksPatch = {
    boardIDs: string[]
    boardPatches: BoardPatch[]
    blockIDs: string[]
    blockPatches: BlockPatch[]
}

type PropertyTypeEnum = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'multiPerson' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy' | 'unknown'

interface IPropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    id: string
    name: string
    type: PropertyTypeEnum
    options: IPropertyOption[]
}

function createBoard(board?: Board): Board {
    const now = Date.now()
    let cardProperties: IPropertyTemplate[] = []
    const selectProperties = cardProperties.find((o) => o.type === 'select')
    if (!selectProperties) {
        const property: IPropertyTemplate = {
            id: Utils.createGuid(IDType.BlockID),
            name: 'Status',
            type: 'select',
            options: [],
        }
        cardProperties.push(property)
    }

    if (board?.cardProperties) {
        // Deep clone of card properties and their options
        cardProperties = board?.cardProperties.map((o: IPropertyTemplate) => {
            return {
                id: o.id,
                name: o.name,
                type: o.type,
                options: o.options ? o.options.map((option) => ({...option})) : [],
            }
        })
    }

    return {
        id: board?.id || Utils.createGuid(IDType.Board),
        teamId: board?.teamId || '',
        channelId: board?.channelId || '',
        createdBy: board?.createdBy || '',
        modifiedBy: board?.modifiedBy || '',
        type: board?.type || BoardTypePrivate,
        minimumRole: board?.minimumRole || MemberRole.None,
        title: board?.title || '',
        description: board?.description || '',
        icon: board?.icon || '',
        showDescription: board?.showDescription || false,
        isTemplate: board?.isTemplate || false,
        templateVersion: board?.templateVersion || 0,
        properties: board?.properties || {},
        cardProperties,
        createAt: board?.createAt || now,
        updateAt: board?.updateAt || now,
        deleteAt: board?.deleteAt || 0,
    }
}

type BoardGroup = {
    option: IPropertyOption
    cards: Card[]
}

// getPropertiesDifference returns a list of the property IDs that are
// contained in propsA but are not contained in propsB
function getPropertiesDifference(propsA: IPropertyTemplate[], propsB: IPropertyTemplate[]): string[] {
    const diff: string[] = []
    propsA.forEach((val) => {
        if (!propsB.find((p) => p.id === val.id)) {
            diff.push(val.id)
        }
    })

    return diff
}

// isPropertyEqual checks that both the contents of the property and
// its options are equal
function isPropertyEqual(propA: IPropertyTemplate, propB: IPropertyTemplate): boolean {
    for (const val of Object.keys(propA)) {
        if (val !== 'options' && (propA as any)[val] !== (propB as any)[val]) {
            return false
        }
    }

    if (propA.options.length !== propB.options.length) {
        return false
    }

    for (const opt of propA.options) {
        const optionB = propB.options.find((o) => o.id === opt.id)
        if (!optionB) {
            return false
        }

        for (const val of Object.keys(opt)) {
            if ((opt as any)[val] !== (optionB as any)[val]) {
                return false
            }
        }
    }

    return true
}

// createCardPropertiesPatches creates two BoardPatch instances, one that
// contains the delta to update the board cardProperties and another one for
// the undo action, in case it happens
function createCardPropertiesPatches(newCardProperties: IPropertyTemplate[], oldCardProperties: IPropertyTemplate[]): BoardPatch[] {
    const newDeletedCardProperties = getPropertiesDifference(newCardProperties, oldCardProperties)
    const oldDeletedCardProperties = getPropertiesDifference(oldCardProperties, newCardProperties)
    const newUpdatedCardProperties: IPropertyTemplate[] = []
    newCardProperties.forEach((val) => {
        const oldCardProperty = oldCardProperties.find((o) => o.id === val.id)
        if (!oldCardProperty || !isPropertyEqual(val, oldCardProperty)) {
            newUpdatedCardProperties.push(val)
        }
    })
    const oldUpdatedCardProperties: IPropertyTemplate[] = []
    oldCardProperties.forEach((val) => {
        const newCardProperty = newCardProperties.find((o) => o.id === val.id)
        if (!newCardProperty || !isPropertyEqual(val, newCardProperty)) {
            oldUpdatedCardProperties.push(val)
        }
    })

    return [
        {
            updatedCardProperties: newUpdatedCardProperties,
            deletedCardProperties: oldDeletedCardProperties,
        },
        {
            updatedCardProperties: oldUpdatedCardProperties,
            deletedCardProperties: newDeletedCardProperties,
        },
    ]
}

// createPatchesFromBoards creates two BoardPatch instances, one that
// contains the delta to update the board and another one for the undo
// action, in case it happens
function createPatchesFromBoards(newBoard: Board, oldBoard: Board): BoardPatch[] {
    const newDeletedProperties = difference(Object.keys(newBoard.properties || {}), Object.keys(oldBoard.properties || {}))

    const newUpdatedProperties: Record<string, any> = {}
    Object.keys(newBoard.properties || {}).forEach((val) => {
        if (oldBoard.properties[val] !== newBoard.properties[val]) {
            newUpdatedProperties[val] = newBoard.properties[val]
        }
    })

    const newData: Record<string, any> = {}
    Object.keys(newBoard).forEach((val) => {
        if (val !== 'properties' &&
            val !== 'cardProperties' &&
            (oldBoard as any)[val] !== (newBoard as any)[val]) {
            newData[val] = (newBoard as any)[val]
        }
    })

    const oldDeletedProperties = difference(Object.keys(oldBoard.properties || {}), Object.keys(newBoard.properties || {}))

    const oldUpdatedProperties: Record<string, any> = {}
    Object.keys(oldBoard.properties || {}).forEach((val) => {
        if (newBoard.properties[val] !== oldBoard.properties[val]) {
            oldUpdatedProperties[val] = oldBoard.properties[val]
        }
    })

    const oldData: Record<string, any> = {}
    Object.keys(oldBoard).forEach((val) => {
        if (val !== 'properties' &&
            val !== 'cardProperties' &&
            (newBoard as any)[val] !== (oldBoard as any)[val]) {
            oldData[val] = (oldBoard as any)[val]
        }
    })

    const [cardPropertiesPatch, cardPropertiesUndoPatch] = createCardPropertiesPatches(newBoard.cardProperties, oldBoard.cardProperties)

    return [
        {
            ...newData,
            ...cardPropertiesPatch,
            updatedProperties: newUpdatedProperties,
            deletedProperties: oldDeletedProperties,
        },
        {
            ...oldData,
            ...cardPropertiesUndoPatch,
            updatedProperties: oldUpdatedProperties,
            deletedProperties: newDeletedProperties,
        },
    ]
}

function createPatchesFromBoardsAndBlocks(updatedBoard: Board, oldBoard: Board, updatedBlockIDs: string[], updatedBlocks: Block[], oldBlocks: Block[]): BoardsAndBlocksPatch[] {
    const blockUpdatePatches = [] as BlockPatch[]
    const blockUndoPatches = [] as BlockPatch[]
    updatedBlocks.forEach((newBlock, i) => {
        const [updatePatch, undoPatch] = createPatchesFromBlocks(newBlock, oldBlocks[i])
        blockUpdatePatches.push(updatePatch)
        blockUndoPatches.push(undoPatch)
    })

    const [boardUpdatePatch, boardUndoPatch] = createPatchesFromBoards(updatedBoard, oldBoard)

    const updatePatch: BoardsAndBlocksPatch = {
        blockIDs: updatedBlockIDs,
        blockPatches: blockUpdatePatches,
        boardIDs: [updatedBoard.id],
        boardPatches: [boardUpdatePatch],
    }

    const undoPatch: BoardsAndBlocksPatch = {
        blockIDs: updatedBlockIDs,
        blockPatches: blockUndoPatches,
        boardIDs: [updatedBoard.id],
        boardPatches: [boardUndoPatch],
    }

    return [updatePatch, undoPatch]
}

export {
    Board,
    BoardPatch,
    BoardMember,
    BoardsAndBlocks,
    BoardsAndBlocksPatch,
    PropertyTypeEnum,
    IPropertyOption,
    IPropertyTemplate,
    BoardGroup,
    createBoard,
    BoardTypes,
    BoardTypeOpen,
    BoardTypePrivate,
    MemberRole,
    createPatchesFromBoards,
    createPatchesFromBoardsAndBlocks,
    createCardPropertiesPatches,
}
