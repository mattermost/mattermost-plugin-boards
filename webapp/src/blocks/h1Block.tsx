// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ContentBlock} from './contentBlock'
import {Block, createBlock} from './block'

type H1Block = ContentBlock & {
    type: 'h1'
}

function createH1Block(block?: Block): H1Block {
    return {
        ...createBlock(block),
        type: 'h1',
    }
}

export {H1Block, createH1Block}

