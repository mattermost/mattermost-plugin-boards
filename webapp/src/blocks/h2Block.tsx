// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ContentBlock} from './contentBlock'
import {Block, createBlock} from './block'

type H2Block = ContentBlock & {
    type: 'h2'
}

function createH2Block(block?: Block): H2Block {
    return {
        ...createBlock(block),
        type: 'h2',
    }
}

export {H2Block, createH2Block}

