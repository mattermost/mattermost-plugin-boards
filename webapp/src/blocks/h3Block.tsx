// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ContentBlock} from './contentBlock'
import {Block, createBlock} from './block'

type H3Block = ContentBlock & {
    type: 'h3'
}

function createH3Block(block?: Block): H3Block {
    return {
        ...createBlock(block),
        type: 'h3',
    }
}

export {H3Block, createH3Block}

