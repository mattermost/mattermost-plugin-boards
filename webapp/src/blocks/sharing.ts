// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.


interface ISharing {
    id: string
    enabled: boolean
    token: string
    modifiedBy?: string
    updateAt?: number
}

export {ISharing}
