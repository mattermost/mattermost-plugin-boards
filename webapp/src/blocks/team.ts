// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

interface ITeam {
    readonly id: string
    readonly title: string
    readonly signupToken: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    readonly settings: Readonly<Record<string, any>>
    readonly modifiedBy?: string
    readonly updateAt?: number
}

export {ITeam}
