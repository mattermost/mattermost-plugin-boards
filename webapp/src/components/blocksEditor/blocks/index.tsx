// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ContentType} from './types'
import H1 from './h1'
import H2 from './h2'
import H3 from './h3'
import Image from './image'
import Text from './text'
import Divider from './divider'

// import Markdown from './markdown'
import ListItem from './list-item'
import Attachment from './attachment'
import Quote from './quote'
import Video from './video'
import Checkbox from './checkbox'

const blocks: {[key: string]: ContentType} = {}
const blocksByPrefix: {[key: string]: ContentType} = {}
const blocksBySlashCommand: {[key: string]: ContentType} = {}

export function register(contentType: ContentType<any>) {
    blocks[contentType.name] = contentType
    if (contentType.prefix !== '') {
        blocksByPrefix[contentType.prefix] = contentType
    }
    blocksBySlashCommand[contentType.slashCommand] = contentType
}

export function list() {
    return Object.values(blocks)
}

export function get(name: string): ContentType {
    return blocks[name]
}

export function getByPrefix(prefix: string): ContentType {
    return blocksByPrefix[prefix]
}

export function isSubPrefix(text: string): boolean {
    for (const ct of list()) {
        if (ct.prefix !== '' && ct.prefix.startsWith(text)) {
            return true
        }
    }
    return false
}

export function getBySlashCommand(slashCommand: string): ContentType {
    return blocksBySlashCommand[slashCommand]
}

export function getBySlashCommandPrefix(slashCommandPrefix: string): ContentType|null {
    for (const ct of list()) {
        if (ct.slashCommand.startsWith(slashCommandPrefix)) {
            return ct
        }
    }
    return null
}

register(H1)
register(H2)
register(H3)
register(Image)
register(Text)
register(Divider)

// register(Markdown)
register(ListItem)
register(Attachment)
register(Quote)
register(Video)
register(Checkbox)
