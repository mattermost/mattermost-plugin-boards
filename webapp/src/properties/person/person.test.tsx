// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.


import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {render, waitFor} from '@testing-library/react'

import configureStore from 'redux-mock-store'

import {act} from 'react-dom/test-utils'

import userEvent from '@testing-library/user-event'

import {wrapIntl} from '../../testUtils'
import {IPropertyTemplate, Board} from '../../blocks/board'
import {Card} from '../../blocks/card'

import PersonProperty from './property'
import Person from './person'

describe('properties/person', () => {
    const mockStore = configureStore([])
    const state = {
        users: {
            boardUsers: {
                'user-id-1': {
                    id: 'user-id-1',
                    username: 'username-1',
                    email: 'user-1@example.com',
                    props: {},
                    create_at: 1621315184,
                    update_at: 1621315184,
                    delete_at: 0,
                },
            },
        },
        clientConfig: {
            value: {
                teammateNameDisplay: 'username',
            },
        },
    }

    test('not readOnly not existing user', async () => {
        const store = mockStore(state)
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <Person
                    property={new PersonProperty()}
                    propertyValue={'user-id-2'}
                    readOnly={false}
                    showEmptyPlaceholder={false}
                    propertyTemplate={{} as IPropertyTemplate}
                    board={{} as Board}
                    card={{} as Card}
                />
            </ReduxProvider>,
        )

        const renderResult = render(component)
        const container = await waitFor(() => {
            if (!renderResult.container) {
                return Promise.reject(new Error('container not found'))
            }
            return Promise.resolve(renderResult.container)
        })
        expect(container).toMatchSnapshot()
    })

    test('not readonly', async () => {
        const store = mockStore(state)
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <Person
                    property={new PersonProperty()}
                    propertyValue={'user-id-1'}
                    readOnly={false}
                    showEmptyPlaceholder={false}
                    propertyTemplate={{} as IPropertyTemplate}
                    board={{} as Board}
                    card={{} as Card}
                />
            </ReduxProvider>,
        )

        const renderResult = render(component)
        const container = await waitFor(() => {
            if (!renderResult.container) {
                return Promise.reject(new Error('container not found'))
            }
            return Promise.resolve(renderResult.container)
        })
        expect(container).toMatchSnapshot()
    })

    test('not readonly guest user', async () => {
        const store = mockStore({...state, users: {boardUsers: {'user-id-1': {...state.users.boardUsers['user-id-1'], is_guest: true}}}})
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <Person
                    property={new PersonProperty()}
                    propertyValue={'user-id-1'}
                    readOnly={false}
                    showEmptyPlaceholder={false}
                    propertyTemplate={{} as IPropertyTemplate}
                    board={{} as Board}
                    card={{} as Card}
                />
            </ReduxProvider>,
        )

        const renderResult = render(component)
        const container = await waitFor(() => {
            if (!renderResult.container) {
                return Promise.reject(new Error('container not found'))
            }
            return Promise.resolve(renderResult.container)
        })
        expect(container).toMatchSnapshot()
    })

    test('readonly view', async () => {
        const store = mockStore(state)
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <Person
                    property={new PersonProperty()}
                    propertyValue={'user-id-1'}
                    readOnly={true}
                    showEmptyPlaceholder={false}
                    propertyTemplate={{} as IPropertyTemplate}
                    board={{} as Board}
                    card={{} as Card}
                />
            </ReduxProvider>,
        )

        const renderResult = render(component)
        const container = await waitFor(() => {
            if (!renderResult.container) {
                return Promise.reject(new Error('container not found'))
            }
            return Promise.resolve(renderResult.container)
        })
        expect(container).toMatchSnapshot()
    })

    test('user dropdown open', async () => {
        const store = mockStore(state)
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <Person
                    property={new PersonProperty()}
                    propertyValue={'user-id-1'}
                    readOnly={false}
                    showEmptyPlaceholder={false}
                    propertyTemplate={{} as IPropertyTemplate}
                    board={{} as Board}
                    card={{} as Card}
                />
            </ReduxProvider>,
        )

        const renderResult = render(component)
        const container = await waitFor(() => {
            if (!renderResult.container) {
                return Promise.reject(new Error('container not found'))
            }
            return Promise.resolve(renderResult.container)
        })

        if (container) {
            // this is the actual element where the click event triggers
            // opening of the dropdown
            const userProperty = container.querySelector('.Person > div > div:nth-child(1) > div:nth-child(2) > input')
            expect(userProperty).not.toBeNull()

            act(() => {
                userEvent.click(userProperty as Element)
            })
            expect(container).toMatchSnapshot()
        } else {
            throw new Error('container should have been initialized')
        }
    })
})
