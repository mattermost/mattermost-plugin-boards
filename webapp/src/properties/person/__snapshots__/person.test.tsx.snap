// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`properties/person not readOnly not existing user 1`] = `
<div>
  <div
    class="Person octo-propertyvalue css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="react-select__control css-18140j1-Control"
    >
      <div
        class="react-select__value-container css-1gbdvdc-ValueContainer"
      >
        <div
          class="react-select__placeholder css-1jqq78o-placeholder"
          id="react-select-2-placeholder"
        >
          Empty
        </div>
        <div
          class="react-select__input-container css-1p5v8kp-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-2-placeholder"
            aria-expanded="false"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="react-select__input"
            id="react-select-2-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class="react-select__indicator-separator css-1uei4ir-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-uycnsi-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`properties/person not readonly 1`] = `
<div>
  <div
    class="Person octo-propertyvalue css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="react-select__control css-18140j1-Control"
    >
      <div
        class="react-select__value-container react-select__value-container--has-value css-1gbdvdc-ValueContainer"
      >
        <div
          class="react-select__single-value css-qosd1h-singleValue"
        >
          <div
            class="Person-item"
          >
            username-1
          </div>
        </div>
        <div
          class="react-select__input-container css-1p5v8kp-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="react-select__input"
            id="react-select-3-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__clear-indicator css-31haax-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="react-select__indicator-separator css-1uei4ir-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-uycnsi-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`properties/person not readonly guest user 1`] = `
<div>
  <div
    class="Person octo-propertyvalue css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-4-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="react-select__control css-18140j1-Control"
    >
      <div
        class="react-select__value-container react-select__value-container--has-value css-1gbdvdc-ValueContainer"
      >
        <div
          class="react-select__single-value css-qosd1h-singleValue"
        >
          <div
            class="Person-item"
          >
            username-1
            <div
              class="GuestBadge"
            >
              <div
                class="GuestBadge__box"
              >
                Guest
              </div>
            </div>
          </div>
        </div>
        <div
          class="react-select__input-container css-1p5v8kp-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="react-select__input"
            id="react-select-4-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__clear-indicator css-31haax-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="react-select__indicator-separator css-1uei4ir-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-uycnsi-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`properties/person readonly view 1`] = `
<div>
  <div
    class="Person octo-propertyvalue octo-propertyvalue--readonly"
  >
    <div
      class="Person-item"
    >
      username-1
    </div>
  </div>
</div>
`;

exports[`properties/person user dropdown open 1`] = `
<div>
  <div
    class="Person octo-propertyvalue css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-5-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
        option username-1 selected, 1 of 1. 1 result available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class="react-select__control react-select__control--is-focused react-select__control--menu-is-open css-18140j1-Control"
    >
      <div
        class="react-select__value-container react-select__value-container--has-value css-1gbdvdc-ValueContainer"
      >
        <div
          class="react-select__single-value css-qosd1h-singleValue"
        >
          <div
            class="Person-item"
          >
            username-1
          </div>
        </div>
        <div
          class="react-select__input-container css-1p5v8kp-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-controls="react-select-5-listbox"
            aria-expanded="true"
            aria-haspopup="true"
            aria-owns="react-select-5-listbox"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="react-select__input"
            id="react-select-5-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__clear-indicator css-3pqe01-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="react-select__indicator-separator css-1uei4ir-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-zngtjc-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="react-select__menu css-1slvcut-menu"
      id="react-select-5-listbox"
    >
      <div
        class="react-select__menu-list css-1d1qzc4-MenuList"
      >
        <div
          aria-disabled="false"
          class="react-select__option react-select__option--is-focused react-select__option--is-selected css-ad52of-option"
          id="react-select-5-option-0"
          tabindex="-1"
        >
          <div
            class="Person-item"
          >
            username-1
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
